File: integration/zig/src/main_test.zig
const std = @import("std");
const helpers = @import("helpers.zig");

// Import all test modules so the Zig test runner discovers their tests.
comptime {
    _ = @import("p2pkh_test.zig");
    _ = @import("counter_test.zig");
    _ = @import("escrow_test.zig");
    _ = @import("math_demo_test.zig");
    _ = @import("function_patterns_test.zig");
    _ = @import("compile_all_test.zig");
    _ = @import("compile.zig");
    _ = @import("auction_test.zig");
    _ = @import("convergence_proof_test.zig");
    _ = @import("covenant_vault_test.zig");
    _ = @import("ec_isolation_test.zig");
    _ = @import("token_ft_test.zig");
    _ = @import("token_nft_test.zig");
    _ = @import("oracle_price_test.zig");
    _ = @import("schnorr_zkp_test.zig");
    _ = @import("sphincs_wallet_test.zig");
    _ = @import("tic_tac_toe_test.zig");
    _ = @import("post_quantum_wallet_test.zig");
}

test "integration_setup" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not running. Skipping integration tests.", .{});
        std.log.warn("Start with: cd integration && ./regtest.sh start", .{});
        return;
    }

    // Mine initial blocks so coinbase UTXOs mature (100 block maturity).
    const current_height = helpers.getBlockCount(allocator) catch |err| {
        std.log.err("Failed to get block count: {any}", .{err});
        return;
    };

    const target_height: i64 = 101;
    const blocks_needed = target_height - current_height;
    if (blocks_needed > 0) {
        std.log.info("Mining {d} blocks (current height: {d}, target: {d})...", .{ blocks_needed, current_height, target_height });
        helpers.mine(allocator, blocks_needed) catch |err| {
            std.log.err("Failed to mine initial blocks: {any}", .{err});
            return;
        };
    }

    std.log.info("Integration test setup complete. Block height >= {d}", .{target_height});
}

File: integration/go/main_test.go
//go:build integration

package integration

import (
	"fmt"
	"os"
	"testing"

	"runar-integration/helpers"
)

func TestMain(m *testing.M) {
	nodeType := helpers.NodeType()
	fmt.Fprintf(os.Stderr, "Integration tests using node type: %s\n", nodeType)

	if !helpers.IsNodeAvailable() {
		fmt.Fprintln(os.Stderr, "Regtest node not running. Skipping integration tests.")
		if helpers.IsTeranode() {
			fmt.Fprintln(os.Stderr, "Start with: cd integration && ./teranode.sh start")
		} else {
			fmt.Fprintln(os.Stderr, "Start with: cd integration && ./regtest.sh start")
		}
		os.Exit(0)
	}

	helpers.EnsureRegtest()

	// Mine initial blocks so coinbase UTXOs mature (100 block maturity).
	// On Teranode regtest, GenesisActivationHeight=10000 (hardcoded in go-chaincfg),
	// so we need height > 10000 for post-Genesis rules. The teranode.sh start script
	// pre-mines these blocks, but we check and top up if needed.
	currentHeight, err := helpers.GetBlockCount()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to get block count: %v\n", err)
		os.Exit(1)
	}

	targetHeight := 101 // SV Node: genesis at height 1
	if helpers.IsTeranode() {
		targetHeight = 10_101 // regtest genesis at 10000 + 101 for coinbase maturity
	}

	blocksNeeded := targetHeight - currentHeight
	if blocksNeeded > 0 {
		fmt.Fprintf(os.Stderr, "Mining %d blocks (current height: %d, target: %d)...\n", blocksNeeded, currentHeight, targetHeight)
		if err := helpers.Mine(blocksNeeded); err != nil {
			fmt.Fprintf(os.Stderr, "Failed to mine initial blocks: %v\n", err)
			os.Exit(1)
		}
	}

	os.Exit(m.Run())
}
